src/rust/engine/src/scheduler.rs

```diff
@@ -474,6 +476,19 @@ impl Scheduler {
     session.maybe_display_initialize(&self.core.executor, &sender);
     self.execute_helper(request, session, sender);
     let result = loop {
+      match externs::call(&python_signal, &[]) {
```

Comment:
Would be good to move this out into a helper function to keep the loop body small. You might also move the check "after" the call to `recv_timeout` but before the `match`... as the most likely time to receive a signal is while we are waiting, and reacting to it before we render the UI/print_stderr/etc will be the most responsive.